### ISSUE TITLE
Can't create table from the result of a UDF

### ISSUE THREAD

**Issue by linxiaokun528** [greenplum-db/gpdb#7439]
### Greenplum version or build
Greenplum Database 6.0.0-beta.1 build commit:067fb8e43115bb087c872ee3d2269d869430263d

### OS version and uname -a
OS version: centos6
uname -a:Linux gpadmin 4.9.125-linuxkit #1 SMP Fri Sep 7 08:20:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### autoconf options used ( config.status --config )
Don't know. Use the offical rpm package:greenplum-db-6.0.0-beta.1-rhel6-x86_64.rpm

### Installation information ( pg_config )
BINDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/bin
DOCDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/share/doc/postgresql
HTMLDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/share/doc/postgresql
INCLUDEDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/include
PKGINCLUDEDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/include/postgresql
INCLUDEDIR-SERVER = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/include/postgresql/server
LIBDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/lib
PKGLIBDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/lib/postgresql
LOCALEDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/share/locale
MANDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/man
SHAREDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/share/postgresql
SYSCONFDIR = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/etc/postgresql
PGXS = /home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--with-quicklz' '--enable-tap-tests' '--enable-gpperfmon' '--with-gssapi' '--enable-mapreduce' '--enable-orafce' '--enable-orca' '--with-libxml' '--with-pgport=5432' '--with-libedit-preferred' '--with-perl' '--with-python' '--with-includes=/tmp/build/f8c7ee08/gpdb_src/gpAux/ext/rhel6_x86_64/include /tmp/build/f8c7ee08/gpdb_src/gpAux/ext/rhel6_x86_64/include/libxml2' '--with-libraries=/tmp/build/f8c7ee08/gpdb_src/gpAux/ext/rhel6_x86_64/lib' '--with-openssl' '--with-pam' '--with-ldap' '--prefix=/usr/local/greenplum-db-devel' '--mandir=/usr/local/greenplum-db-devel/man' 'CC=gcc -m64 -gdwarf-2 -gstrict-dwarf' 'CFLAGS=-m64 -gdwarf-2 -gstrict-dwarf -O3 -fargument-noalias-global -fno-omit-frame-pointer -g'
CC = gcc -m64 -gdwarf-2 -gstrict-dwarf
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2 -I/tmp/build/f8c7ee08/gpdb_src/gpAux/ext/rhel6_x86_64/include -I/home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -m64 -gdwarf-2 -gstrict-dwarf -O3 -fargument-noalias-global -fno-omit-frame-pointer -g -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/include
CFLAGS_SL = -fPIC
LDFLAGS = -L/tmp/build/f8c7ee08/gpdb_src/gpAux/ext/rhel6_x86_64/lib -Wl,--as-needed -Wl,-rpath,'/usr/local/greenplum-db-devel/lib',--enable-new-dtags -L/home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/lib
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lxml2 -lpam -lrt -lyaml -lgssapi_krb5 -lquicklz -lzstd -lrt -lcrypt -ldl -lm -L/home/gpadmin/workspace/gptext/gpdb-solr/.gpdb6_tmp/greenplum-db-6.0.0-beta.1/lib
VERSION = PostgreSQL 9.4.20

### Expected behavior
Executing the following SQL should be successful:
CREATE TABLE example(a int, b text) DISTRIBUTED by (a);
insert into example value(1,'1'),(2,'2');
CREATE FUNCTION multiset_2(a anytable) RETURNS TABLE(a int, b text) AS '/home/gpadmin/workspace/gpdb/src/test/regress/regress.so', 'multiset_example' LANGUAGE C READS SQL DATA;        --Need the regress.so file
create function scalar_3() RETURNS table(a int, b text)
as $$ select * from multiset_2( TABLE( SELECT * from example) ) order by a, b $$ LANGUAGE SQL;
create table test_b as select * from scalar_3();

### Actual behavior
When executing the last sql command, we have the following error message:
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column(s) named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
ERROR:  function cannot execute on a QE slice because it accesses relation "public.example"  (entry db 127.0.0.1:15432 pid=24239)
CONTEXT:  SQL function "scalar_3" during startup

### Step to reproduce the behavior
Put file regress.so to /home/gpadmin/workspace/gpdb/src/test/regress/regress.so
Execute the following sql command:
CREATE TABLE example(a int, b text) DISTRIBUTED by (a);
insert into example value(1,'1'),(2,'2');
CREATE FUNCTION multiset_2(a anytable) RETURNS TABLE(a int, b text) AS '/home/gpadmin/workspace/gpdb/src/test/regress/regress.so', 'multiset_example' LANGUAGE C READS SQL DATA;        --Need the regress.so file
create function scalar_3() RETURNS table(a int, b text)
as $$ select * from multiset_2( TABLE( SELECT * from example) ) order by a, b $$ LANGUAGE SQL;
create table test_b as select * from scalar_3();

**Comment by berlin-ab** [comment]
How could this be done without regress.so? What's an example function that we could use to replicate the issue w/o additional `.so` files?

**Comment by asimrp** [comment]
Thanks for the crisp repro.  CTAS plan needs to redistribute the tuples returned by the select clause so that they can be inserted into the table.  This is done by a new backend process called entrydb, represented by RedistributeMotion 1:3 node in the explain plan:
```
explain create table test_b as select * from scalar_3();
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause. Creating a NULL policy entry.
                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=0)
   ->  Result  (cost=0.00..20.88 rows=334 width=12)
         ->  Redistribute Motion 1:3  (slice1)  (cost=0.00..0.04 rows=1000 width=12)
               ->  Function Scan on scalar_3  (cost=0.00..0.01 rows=334 width=12)
 Optimizer: PQO version 3.35.0
```

Accessing distributed tables from QE segments is prohibited in Greenplum.  I'm wondering if entrydb QE can be exempted from this restriction because it is a special QE that runs on master.

**Comment by d** [comment]
> I'm wondering if entrydb QE can be exempted from this restriction because it is a special QE that runs on master.

Yes but mostly no. Only in simple cases where we don't already have a busy dispatcher on QD. I imagine for a CTAS, marking the function as `EXECUTE ON MASTER` won't help because CTAS itself requires a dispatch

**Comment by d** [comment]
This is a known limit in Greenplum's dispatcher:

If a function requires a dispatch (e.g. requires any kind of motion), then the plan including the function call had better:
1. Place the call on QD
1. Require no dispatching of the main plan (i.e. no motions)
